features/123-chat-push-notification/api.md

```diff
@@ -0,0 +1,158 @@
+# Push Notification in Chat
+## Feature Overview
+
+This feature enables developer to send customized push notification via cloud function deployment. The approach is to provide hooks in chat conversation including
+
+* Messages
+  * after\_message\_sent
+  * after\_message\_updated
+  * after\_message\_deleted
+* Typing
+  * typing\_started
+* Conversation
+  * after\_conversation\_created
+  * after\_conversation\_deleted
+  * after\_conversation\_updated
+  * after\_users\_entered\_conversation
+  * after\_users\_leave\_conversation
+
+In addition, the current JS push notification code snippet is too verbose and can be simplified.
+
+### Task
+- Refactor JS SDK Push Notification Function so that it can be called by developer with less code.
+	- Original Code:
+
+	```javascript
+	var skygear = require('skygear');
+	skygear.push.sendToDevice(
+	  ['2aa4af2a-699a-4e43-8d67-7598757fc7ed'], // Device IDs
+	  {
+	    'apns': {
+	        'aps': {
+	            'alert': {
+	                'title': title,
+	                'body': message,
+	            }
+	        },
+	        'from': 'skygear',
+	        'operation': 'notification',
+	    },
+	    'gcm': {
+	         'notification': {
+	              'title': title,
+	              'body': message,
+	          }
+	    },
+		}
+	);
+	```
+
+	- new Code:
+
+	```javascript
+	var skygear = require('skygear');
+	skygear.push.sendToUser(
+	  [userBen, userRick],
+	  {
+         'title': title,
+         'body': message
+	  }
+	);
+	```
+- Create chat hook annotation in chat SDK
+- Enable chat plugin to call pre-defined hook
+
+### Sample Codes
+
+- Sample Javascript Cloud Function
+
+```javascript
+skygearChat.afterMessageSent((message, conversation) => {
+  skygear.push.sendToUser(conversation.participants, {'title': 'Title', 'body': 'Hello World'})
+})
+```
+
+- Python Cloud Function
+
+```python
+
+@chat.after_message_sent
+def after_messange_send_handler(message, conversation):
+   user_ids = [p.id.key for p in conversation.participants]
+   notification = {...}
+	skygear.action.push_users(container, user_ids, notification)
+```
+
+### List of APIs
+
+- Javascript
+	- `skygearChat.afterMessageSent((message, conversation) => {})`
```

Comment:
In 90% of the use-case, we are sending the push notification to receiving user. So probably `users` should appear in the function parameter?

Reply:
Which user you are talking about?
message.owner or conversation.participants ? In either case, the user can be fetched from function argument.

Reply:
According to your argument, we don't need `conversation` too. Since `conversation` can retrieve from `message.conversation`.

I think explicit function parameter will be helpful for user to do things without checking the full guide and documentation.

Reply:
Agree. I will update the arguments.